### DISCUSSION TITLE
[SOLVED] What are reasons of using Super server mode by default?

### DISCUSSION THREAD

**Discussion by gsbelarus** [FirebirdSQL/firebird#8071] [Q&A]
**UPDATED** after changes were made by Vlad:

| Mode | Page size | Buffers | PW | switches | Time | Note |
|-------|-----------|--------|-----|------|------|------|
| Classic | 32768 | 500 | 16 |  | 2:08:00 |FileSystemCacheThreshold = 100M, TempCacheLimit = 32000M |
| Super | 8192 | 5_000_000 | 16 |  | 2:24:00| FileSystemCacheThreshold = 100M, TempCacheLimit = 32000M |

I have conducted a test:

1. WI-V5.0.1.1375 Firebird 5.0 65d5aa3
1. OS Windows Server 2019
1. Database size 334 GB
2. BK file size 119 GB
3. Server with 128 GB RAM, 20 cores CPU and fast NVMe drive

### firebird.conf

Until overridden by the values specified in the table below, the following settings were applied to the firebird.conf.:

```
MaxParallelWorkers = 64
TempBlockSize = 1M
TempCacheLimit = 16384M
MaxIdentifierByteLength = 31
DataTypeCompatibility = 2.5
WireCrypt = Disabled
RemoteServicePort = 3056
IpcName = FIREBIRD5
AuthServer = Legacy_Auth
AuthClient = Legacy_Auth
UserManager = Legacy_UserManager
```

### Command

```
C:\Program Files\FB5>gbak -r "G:\XXXXXXXX.bk" "G:\XXXXXXXX.fdb" -user sysdba -pas XXXXXXXX -se localhost/3056:service_mgr -p NNNN -bu MMMM -par VVVV
```

### Results

| Mode | Page size | Buffers | PW | switches | Time | Note |
|-------|-----------|--------|-----|------|------|------|
| Classic | 32768 | 500 | 16 |  | 2:07:00 |FileSystemCacheThreshold = 100M, TempCacheLimit = 32000M |
| Classic | 16384 | 500 | 18 |  | 2:13:00 |FileSystemCacheThreshold = 100M, TempCacheLimit = 32000M |
| Classic | 8192 | 500 | 16 |  | 2:20:00 | |
| Classic | 8192 | 500 | 16 |  | 2:22:00 |FileSystemCacheThreshold = 100M, TempCacheLimit = 32000M |
| Super | 8192 | 500 | 16 |  | 5:20:00| FileSystemCacheThreshold = 100M, TempCacheLimit = 32000M |
| Super | 8192 | 5_000_000 | 16 |  | 5:30:00| FileSystemCacheThreshold = 100M, TempCacheLimit = 32000M |
| Super | 8192 | 4_000_000 | 1 | -par 16 | 5:45:00 | |
| Super | 16384 | 100_000 | 1 | -par 16  |5:54:00 | FileSystemCacheThreshold = 100M, TempCacheLimit = 32000M|
| Super | 8192 | 500 | 16 |  | 6:00:00 | |
| Super | 8192 | 5_000_000 | 1 |-par 16  |6:03:00 | FileSystemCacheThreshold = 100M, TempCacheLimit = 32000M |
| Super | 16384 | 100_000 | 1 | -par 16 | 6:20:00| |

* PW -- Parallel Workers parameter in the `firebird.conf`

### Conclusion

The Super Server is not only **less robust and reliable** than the Classic Server, but it also performs **much worse**. So, why do we offer it as the default mode? This greatly harms Firebird because many users do not bother to make adjustments in the Firebird configuration file, and therefore end up with a worse experience.

**Comment by alexeykovyazin** [comment]
Hello Andrei,

First of all, you did not conducted a test: you just create useless ticket which distracts attention of Firebird developers from the real problems.

In your ticket, there is lack of very important details to reproduce a test:

1) Full Firebird Configuration files used in testing, including full Firebird version
2) Full restore log with enabled statistics (switches -ST TDRW -y logname.txt)
3) Details of Operating System used
4) Details of structure of database tested - output of gstat -r ... (it would be good also to have some justification why 8K page size is used, which is totally inappropriate for this database size).
5) Complete test plan - the selection of page buffers = 500 for SuperServer is absolutely inappropriate.  There are other very important parameters - enable/disable file cache of operating system? TempCacheLimit size?
6) Optionally, it would be good to see results of simple INSERT?UPDATE/DELETE test https://ib-aid.com/en/simple-insert-update-delete-test-for-firebird/ with a lot of test results to compare.

If you are now aware about principles how to calculate the configuration, please use Configuration Calculator For Firebird to create optimized configurations for the specified hardware, for Super and Classic/SuperClassic.

If you really want to bring the value to Firebird development, please repeat the tests with optimized configurations, supply all requested information, and submit it for investigation. I will bet a bottle of best zubrowka that SuperServer will win.

Regards,
Alexey Kovyazin
President Of Firebird Foundation

**Comment by sim1984** [comment]
First, you should familiarize yourself with how the `ParallelWorkers` parameter works in SuperServer and in Classic. You know that it is per-process, right?

For the purity of the experiment, it is better to return this parameter to the default value of 1. And compare using the `-par` switch. Well, they also told you about the page cache, it must be meaningfully selected taking into account other parameters, and not just tried two extreme values.

**Comment by mrotteveel** [comment]
I have converted this to a discussion, because as Alexey points out, this wasn't really an issue/ticket.

**Comment by gsbelarus** [comment]
> First, you should familiarize yourself with how the `ParallelWorkers` parameter works in SuperServer and in Classic. You know that it is per-process, right?
> 
> For the purity of the experiment, it is better to return this parameter to the default value of 1. And compare using the `-par` switch. Well, they also told you about the page cache, it must be meaningfully selected taking into account other parameters, and not just tried two extreme values.

What buffers number isn't "extreme" in this case? 100_000? 10_000? I thought the reason to use SuperServer is to have buffer large enough to fit all frequent data. 

We tried 16K, 32K pages as soon as they were introduced and never had a positive effect on our real life cases.

But, any way, I will try it right now.

So, 

Parallel Workers = 1
-par 16
-page_size 16384
-buffers 100_000

Let's see how it will behave.

**Reply by hvlad** [reply]
Set TempCacheLimit  to 1G at least (for SS only!).
And use verbose log with stats (point 2 from Alexey mesage)
Also, consider to use or not file system caching (currenlty you going to not use it)

**Comment by aafemt** [comment]
If database restore is a main task for your server then Classic may be a right choice but usually Firebird is employed for much wider variety of operations.

**Comment by gsbelarus** [comment]
I see. I explain why backup/restore is a crucial point for us. The database is used with large enterprise, where we are allowed only 2-3 maintenance windows per year of 8-12 hours each. Every hour of unexpected downtime could cost USD 100K+. It is obvious that management is very concerned about how fast operability could be restored in case of unexpected events. 

We have fbackup every two hours on a separate server. But, we are considering every scenario and in our history there were cases when database got corrupted and we needed to gfix it and then do full backup restore.

Once it was because of a bug in FB 3.0 SuperServer and we never tried and wouldn't ever try SuperServer again, because we had great problems that time. When there are up to 400 simultaneous connections it is hard task to say 400 workers that they need to reenter some data or recheck their whole daily work.

PS: BTW, 2:20 for the database restoring time we get with Classic looks great! Now it takes around 8 hours with FB 3.0.9.

**Reply by aafemt** [reply]
For such critical database you should consider hot standby instead.

**Reply by gsbelarus** [reply]
agree, but we are on FB 3 and replication was introduced in FB 4? Am I right?

**Reply by aafemt** [reply]
No. Replication is available in HQBird staring from 2.5 IIRC.

**Reply by alexeykovyazin** [reply]
Just to clarify, downtime cost is 1000000 USD/hour, but you cannot invest neither in professional tools nor into support?

**Comment by gsbelarus** [comment]
@hvlad I have turned on file system cache for the SuperServer, but result is the same. 

I used switch -st, but there aren't statistics in the output file. There is a content:
```
gbak: WARNING:function BIN_AND is not defined
gbak: WARNING:    module name or entrypoint could not be found
gbak: WARNING:    function BIN_OR is not defined
gbak: WARNING:    module name or entrypoint could not be found
gbak: WARNING:    function ABS is not defined
gbak: WARNING:    module name or entrypoint could not be found
gbak: WARNING:function ABS is not defined
gbak: WARNING:    module name or entrypoint could not be found
gbak: WARNING:    function BIN_AND is not defined
gbak: WARNING:    module name or entrypoint could not be found
```
is `-st` supposed to work with `-se`?

The whole command line is here:
```
gbak -r "G:\XXXXXX.bk" "G:\XXXXXX.fdb" -user sysdba -pas XXXXXXX -se localhost/3056:service_mgr -p 16384 -bu 100000 -par 16 -ST TDRW -y "G:\XXXXXXXX.txt"
```

**Reply by hvlad** [reply]
Statistics is a part of verbose output, you missed the -verbose switch.
Also, to make any conclusion we need to compare verbose outputs of 'fast' and 'slow' cases, i.e. stats of just one of them makes no much meanings.

**Reply by gsbelarus** [reply]
ok. I will try to recollect the data during the weekend.

**Reply by aafemt** [reply]
And while recollecting data perform restore in Super mode first and in Classic mode then. To eliminate possibility that your SSD is busy with housekeeping.

**Comment by sim1984** [comment]
> Max parallel workers = 64

Seriously? Is this parameter written exactly like that?
Maybe after all

```
MaxParallelWorkers = 64
```

???

**Reply by gsbelarus** [reply]
![image](https://github.com/FirebirdSQL/firebird/assets/5175764/12c3fbcd-b004-4767-8ff2-682b6b4c32da)

**Reply by gsbelarus** [reply]
Above I put an excerpt in a readable format. Of course, every parameter named in a proper way in the firebird.conf file.

**Reply by sim1984** [reply]
I don't really believe it. Because my measurements do not correlate with yours in any way. My multi-threaded SS restores the database at about the same speed as the classic one.

**Comment by alexeykovyazin** [comment]
Hello All,

I would recommend for everybody stop discussing until the all requested information will be provided, in order to have a productive discussion.

Regards,
Alexey

**Comment by gsbelarus** [comment]
I'm unclear on why it's necessary to limit the number of Parallel Workers to 1 when utilizing the Super Server. This seems counterintuitive and warrants clarification within the firebird.conf settings.

**Reply by hvlad** [reply]
It is not required

**Comment by gsbelarus** [comment]
So, I'm going to do two tests with a verbose output.

Classic
page 8192
buffers 500
parallel workers 16

Super
page 8192
buffers 200_000
parallel workers 1, switch -par 16 in the gbak's command line 

All other parameters are equal. File system cache is enabled. Temp cache limit is set to 32GB.

Any objections?

**Reply by aafemt** [reply]
As I said: test Super first and Classic then. Number of buffers also must be the same for proper testing.

**Reply by hvlad** [reply]
Use the same value of `ParallelWorkers` and the same command line for each run.
No need to add more confusion that we already have :)
As for file system cache - set `UseFileSystemCache = true` and leave `FileSystemCacheThreshold` commented out.

**Comment by hvlad** [comment]
> Number of buffers also must be the same for proper testing.

?????

**Reply by aafemt** [reply]
During restore only one attachment is active so shared page cache in Super has no difference from Classic, right? There is no reason to perform experiments with different parameters.

**Reply by hvlad** [reply]
It is parallel restore with many parallel attachments

**Reply by aafemt** [reply]
Do these parallel attachments use separate data caches in Classic? I had impression that in the case of internal attachments (not only restore, but sweep, GC etc) a single classic process works as "mini-super" sharing the cache between these attachments and not spawn separate classic processes for them.

**Reply by mrotteveel** [reply]
@aafemt You're missing that gbak itself creates multiple attachments when restoring data, which means in the case of Classic, each attachment has its own cache. Only when the indexes are built is one attachment used which then uses multiple threads for rebuilding.

**Reply by aafemt** [reply]
@mrotteveel  Are you sure that it is done during **restore** and not backup only?

**Reply by mrotteveel** [reply]
Yes, unless https://github.com/FirebirdSQL/firebird/blob/v5.0-release/doc/README.gbak is wrong

**Reply by hvlad** [reply]
@aafemt you have wrong impressions.

**Comment by gsbelarus** [comment]
Now, I get confused. So, I'm going to set these values in firebird.conf:
```
UseFileSystemCache = true
#FileSystemCacheThreshold = 100M
ParallelWorkers = 16
MaxParallelWorkers = 64
```

and use the same command line for SuperServer and Classic testing. 

I'll use page 8192 (As I have tested above there is no significant difference between 8192 and 16384, so I stick with what we use for last 25 years).

I'll use 500 buffers.

For both runs, I'll grab the stats.

**Reply by hvlad** [reply]
500 buffers for SS is a nonsense

**Reply by gsbelarus** [reply]
I just wnat to use the same value for Classic and SuperServer. Or should I use different values? Say, 500 here and 50_000 there?

**Reply by gsbelarus** [reply]
I could set even more, but default threshold is 64K and this would turn off file system caching.

**Reply by mrotteveel** [reply]
The problem is that given gbak uses multiple attachments *if* you specify the [`-parallel <worker-count>`](https://firebirdsql.org/file/documentation/html/en/firebirddocs/gbak/firebird-gbak.html#gbak-cmdline-parallel) option, then in the case of Classic, you will have a separate buffer for each individual connection, while SuperServer will share a single buffer for all connections. On the one hand, that means that Classic can do with a smaller buffer, because there will be less trashing of the cache by different connections requiring different pages to be cached causing more cache evictions if the buffer is relatively small, and on the other hand, using a too large buffer with Classic will require more memory.

**Reply by mrotteveel** [reply]
> I could set even more, but default threshold is 64K and this would turn off file system caching.

If `UseFileSystemCache` is set to `true` explicitly, then `FileSystemCacheThreshold` is ignored (the docs are unclear about the effects if neither is set explicitly, but I think that `UseFileSystemCache` is inferred to true, and `FileSystemCacheThreshold` is then ignored).

**Reply by gsbelarus** [reply]
ok. I will use both `UseFileSystemCache=true` and `FileSystemCacheThreshold=100M` it should enable file system cache in every possible scenario.

**Reply by hvlad** [reply]
If you use 500 buffers for CS, then you may think about 500*par buffers for SS for more-or-less fair comparison.
Actually, restore doesn't require too big buffer cache *if* file system cache is used.

**Reply by hvlad** [reply]
@gsbelarus you could also consider the moon phase...

**Reply by gsbelarus** [reply]
well, I'll do three runs

Classic 500 pages
Super 500 pages
Super 50_000 pages

**Comment by reevespaul** [comment]
On Fri, 05 Apr 2024 04:43:24 -0700
Mark Rotteveel ***@***.***> wrote:

> > I could set even more, but default threshold is 64K and this would turn
> > off file system caching.  
> 
> If `UseFileSystemCache` is set to `true` explicitly, then
> `FileSystemCacheThreshold` is ignored (the docs are unclear about the
> effects if neither is set explicitly, but I think that `UseFileSystemCache`
> is inferred to true, and `FileSystemCacheThreshold` is then ignored).
> 

No - it is the other way round. If UseFileSystemCache is commented and
DBCachePages exceeds FileSystemCacheThreshold then the file system cache is
disabled. This goes against the expectation that the uncommented value is the
default.


Paul
-- 
Paul Reeves
https://www.ibphoenix.com
Supporting users of Firebird

**Comment by hvlad** [comment]
Firebird 5.0.1.1738 x64
Windows 10 
RAM 32GB, CPU i7-12700
SSD Samsung 980 Pro (NVMe)
Backup size 60GB
Database size 87GB

firebird.conf
UseFileSystemCache = true
TempCacheLimit = 4G
MaxParallelWorkers = 64

ServerMode = Super
gbak -se inet://service_mgr -rep xxx.fbk xxx.FDB -verbint 1000000 -st tdrw -y xxx.log -bu 100000 -par 16
769 sec

ServerMode = Classic
gbak -se inet://service_mgr -rep xxx.fbk xxx.FDB -verbint 1000000 -st tdrw -y xxx.log -bu 1000 -par 16
876 sec

**Reply by hvlad** [reply]
It is the biggest backup I have at hand

**Comment by gsbelarus** [comment]
So, I did five additional runs on the same machine. Before every run, I had the service restarted. The `firebird.conf` looks like:

```
	Line  190: UdfAccess = Restrict UDF
	Line  286: UseFileSystemCache = true
	Line  313: FileSystemCacheThreshold = 100M
	Line  405: TempBlockSize = 1M
	Line  419: TempCacheLimit = 32000M
	Line  433: MaxIdentifierByteLength = 31
	Line  489: AuthServer = Legacy_Auth
	Line  494: AuthClient = Legacy_Auth #Windows clients
	Line  508: UserManager = Legacy_UserManager
	Line  743: DataTypeCompatibility = 2.5
	Line  771: WireCrypt = Disabled
	Line  844: RemoteServicePort = 3056
	Line 1073: MaxParallelWorkers = 64
	Line 1085: ParallelWorkers = 16
	Line 1136: IpcName = FIREBIRD5
```	

All other parameters are commented, i.e., have default values.

I have put results in the table at the beginning of the thread. Duplicating them here:
```
Classic, 500 pages, 32768 page size: 2:07:00
Classic, 500 pages, 8192 page size: 2:22:00
Super, 500 pages, 8192 page size, 5:20:00
Super, 5_000_000 pages, 8192 page size, 5:30:00
Super, 5_000_000 pages. 8192 page size, PW = 1, -par 16, 6:03:00
```

The statistics per run:

[classic_500.txt](https://github.com/FirebirdSQL/firebird/files/14898342/classic_500.txt)
[classic_500_page_32768.txt](https://github.com/FirebirdSQL/firebird/files/14898344/classic_500_page_32768.txt)
[super_500.txt](https://github.com/FirebirdSQL/firebird/files/14898347/super_500.txt)
[super_5000000.txt](https://github.com/FirebirdSQL/firebird/files/14898346/super_5000000.txt)
[super_5000000_pw_1_par_16.txt](https://github.com/FirebirdSQL/firebird/files/14898348/super_5000000_pw_1_par_16.txt)

**Reply by hvlad** [reply]
Thanks. 
Stats clearly shows some performance problem of SS when building FK's.
For example, compare classic_500.txt and super_5000000.txt:

|   | classic_500, sec | super_5000000, sec |
| --- | --- | --- |
| Data load | 1301 | 1163 | 
| Non-FK indices | 767 | 779 | 
| FK indices | 6423 | 17859 |

**Reply by gsbelarus** [reply]
This is the same database I promised to send. As soon, as I obfuscate it to some degree...

**Reply by hvlad** [reply]
Good, waiting for it.

BTW, not all FK's builds too slow by SS. 
For example: 
FK's on tables INV_MOVEMENT (236'375'561 recs) and INV_BALANCE (96'491'226 recs) is slower, 
while FK's on table INV_CARD (141'776'019 rec) mostly same or faster by SS.

**Reply by gsbelarus** [reply]
the difference lays here:
```
classic:
gbak: 5298.584 162.501 1191118 232780     activating and creating deferred index RDB$FOREIGN211
gbak: 5384.348 85.763 272859 200118     activating and creating deferred index RDB$FOREIGN212
gbak: 5461.228 76.880 272758 200028     activating and creating deferred index RDB$FOREIGN214
gbak: 5587.787 126.559 1690293 215564     activating and creating deferred index RDB$FOREIGN215
gbak: 5706.437 118.650 1383436 206979     activating and creating deferred index RDB$FOREIGN216
gbak: 5790.444 84.006 274880 200174     activating and creating deferred index USR$FKAC_ENTRY11
gbak: 5840.305 49.860 428911 170622     activating and creating deferred index USR$FKAC_ENTRY12
gbak: 5918.783 78.477 311651 189176     activating and creating deferred index USR$FKAC_ENTRY13
gbak: 5996.370 77.587 1190997 179534     activating and creating deferred index USR$FKAC_ENTRY14
gbak: 6057.043 60.672 491528 175531     activating and creating deferred index USR$FKAC_ENTRY15
gbak: 6095.668 38.624 274400 166292     activating and creating deferred index USR$FKAC_ENTRY16
gbak: 6138.028 42.360 336069 167319     activating and creating deferred index USR$FKAC_ENTRY127
gbak: 6185.957 47.929 605666 167945     activating and creating deferred index USR$FKAC_ENTRY128
gbak: 6224.512 38.554 276028 166293     activating and creating deferred index USR$FKAC_ENTRY173
gbak: 6281.710 57.197 269206 177954     activating and creating deferred index USR$FKAC_ENTRY193
gbak: 6331.902 50.192 271131 175692     activating and creating deferred index USR$FKAC_ENTRY795
gbak: 6371.261 39.358 282874 166325     activating and creating deferred index USR$FKAC_ENTRY796
gbak: 6429.519 58.258 277848 180844     activating and creating deferred index USR$FKAC_ENTRY797
```
and
```
super:
gbak: 10137.355 533.164 386767 232327     activating and creating deferred index RDB$FOREIGN211
gbak: 10783.832 646.476 209337 199729     activating and creating deferred index RDB$FOREIGN212
gbak: 11429.842 646.010      5 199639     activating and creating deferred index RDB$FOREIGN214
                ^^^^^  10 times slower
gbak: 11995.068 565.225 376663 215142     activating and creating deferred index RDB$FOREIGN215
gbak: 12617.362 622.294 407166 206575     activating and creating deferred index RDB$FOREIGN216
gbak: 13260.631 643.269 231301 199783     activating and creating deferred index USR$FKAC_ENTRY11
gbak: 13365.432 104.800    184 170290     activating and creating deferred index USR$FKAC_ENTRY12
gbak: 13819.992 454.560      3 188810     activating and creating deferred index USR$FKAC_ENTRY13
gbak: 13950.022 130.029 374089 179183     activating and creating deferred index USR$FKAC_ENTRY14
gbak: 14147.249 197.227 220007 175189     activating and creating deferred index USR$FKAC_ENTRY15
gbak: 14182.634 35.384      0 165970     activating and creating deferred index USR$FKAC_ENTRY16
                ^^^^^ faster than CS!
gbak: 14222.914 40.279    286 166993     activating and creating deferred index USR$FKAC_ENTRY127
gbak: 14280.691 57.777   7415 167618     activating and creating deferred index USR$FKAC_ENTRY128
gbak: 14316.204 35.512      1 165968     activating and creating deferred index USR$FKAC_ENTRY173
gbak: 14555.211 239.006      6 177608     activating and creating deferred index USR$FKAC_ENTRY193
gbak: 14752.099 196.888      5 175349     activating and creating deferred index USR$FKAC_ENTRY795
gbak: 14787.854 35.754    420 166001     activating and creating deferred index USR$FKAC_ENTRY796
                ^^^^ almost the same time
gbak: 15078.718 290.864      7 180489     activating and creating deferred index USR$FKAC_ENTRY797
                ^^^^ 5 times slower
```

**Reply by gsbelarus** [reply]
It is somehow related to the mechanism which is used by the SS server to access pages in the cache. When CS just goes straight to the disk and gets data it is needed, the SS hangs on the inter threads synchronization. That is how I see it.

**Reply by gsbelarus** [reply]
or am I wrong? Do fk names match between CS and SS restored database?

**Reply by hvlad** [reply]
> Do fk names match between CS and SS restored database?

Sure

**Reply by hvlad** [reply]
> It is somehow related to the mechanism which is used by the SS server to access pages in the cache. When CS just goes straight to the disk and gets data it is needed, the SS hangs on the inter threads synchronization. That is how I see it.

It is not as easy as you see it. Else SS would be slow with all indices, or with all FK's, but this is not the case.

**Reply by aafemt** [reply]
It would be good to see definition of slow FK and corresponding PKs to check that data types of fields match. If for example they are VARCHARs with different character sets SS can suffer from serialization of ICU calls.

**Reply by gsbelarus** [reply]
all PKs and FKs in the database are positive 32bit integers.

**Reply by hvlad** [reply]
> This is the same database I promised to send. As soon, as I obfuscate it to some degree...

Reminder...

**Reply by gsbelarus** [reply]
sent instructions how to get the db by email

**Reply by hvlad** [reply]
Thanks, it allows me to identify and fix most significant source of contention in SS that happens when FK is created.
The fix is committed into master now. Some quick results for the one of the "slowest" index:

`alter index RDB$FOREIGN214 active;`

|      | no fix  | with fix |
| --- | --- | --- |
CS  | 166  |  165
SS  | 656  | 179

**Reply by hvlad** [reply]
Committed into v5 also. Please, try next snapshot build.

**Reply by gsbelarus** [reply]
Thanks! I'll try it.

**Reply by sim1984** [reply]
> Thanks! I'll try it.

Are there any new test results after the patch?

**Reply by gsbelarus** [reply]
New results are at the top.
